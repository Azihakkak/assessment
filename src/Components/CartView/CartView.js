import React from "react";

const hundred = 100;
const roundedDecimal = 2;

export default function CartView({ cart, tax }) {
  const taxExcludedTotal = () => {
    let totalPrice = 0;
    Object.values(cart).forEach((item) => {
      totalPrice += +item.price * +item.quantity;
    });
    return (Math.ceil(totalPrice * hundred + Number.EPSILON) / hundred).toFixed(
      roundedDecimal
    );
  };

  const calculateTax = () => {
    const total = taxExcludedTotal();
    const result = (total * tax) / hundred;
    return (Math.ceil(result * hundred + Number.EPSILON) / hundred).toFixed(
      roundedDecimal
    );
  };

  const taxIncludedTotal = () => {
    const total = taxExcludedTotal();
    const tax = calculateTax();
    const result = +total + +tax;
    return result;
  };

  function renderTableData() {
    return Object.entries(cart).map(([key]) => {
      const name = key;
      const quantity = cart[key].quantity;
      const price = cart[key].price;

      return (
        <tr key={name}>
          <td data-testid="name">Item: {name}</td>
          <td data-testid="quantity">Qty: {quantity}</td>
          <td data-testid="price">UnitPrice: ${price}</td>
        </tr>
      );
    });
  }

  return (
    <div>
      <table>
        <tbody>{renderTableData()}</tbody>
      </table>
      <h3>Total Tax: ${calculateTax()}</h3>
      <h3>Total Price: ${taxExcludedTotal()}</h3>
      <h3>Total Price IncludingTax: ${taxIncludedTotal()}</h3>
    </div>
  );
}
