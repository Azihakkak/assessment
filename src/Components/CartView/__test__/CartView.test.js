import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import CartView from "../CartView";

afterEach(cleanup);

it("Should render one item", () => {
  const tree = renderer
    .create(<CartView cart={{ "Shower gel": { quantity: 5, price: 49.99 } }} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it("Should show 2 items in the cart", () => {
  const { getByText, queryAllByTestId } = render(
    <CartView
      cart={{
        "Shower gel": { quantity: 2, price: 49.99 },
        Deodorant: { quantity: 2, price: 99.99 },
      }}
      tax={"12.5"}
    />
  );

  expect(getByText("Item: Shower gel")).toBeInTheDocument();
  expect(getByText("Item: Deodorant")).toBeInTheDocument();
  expect(queryAllByTestId("quantity")[0]).toHaveTextContent("2");
  expect(queryAllByTestId("quantity")[1]).toHaveTextContent("2");
  expect(getByText("UnitPrice: $49.99")).toBeInTheDocument();
  expect(getByText("UnitPrice: $99.99")).toBeInTheDocument();
  expect(getByText("Total Tax: $37.50")).toBeInTheDocument();
  expect(getByText("Total Price IncludingTax: $337.46")).toBeInTheDocument();
});
