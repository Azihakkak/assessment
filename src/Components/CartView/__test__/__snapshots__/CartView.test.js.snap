// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render one item 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          data-testid="name"
        >
          Item: 
          Shower gel
        </td>
        <td
          data-testid="quantity"
        >
          Qty: 
          5
        </td>
        <td
          data-testid="price"
        >
          UnitPrice: $
          49.99
        </td>
      </tr>
    </tbody>
  </table>
  <h3>
    Total Tax: $
    NaN
  </h3>
  <h3>
    Total Price: $
    249.95
  </h3>
  <h3>
    Total Price IncludingTax: $
    NaN
  </h3>
</div>
`;
